libs/core/src/main/java/org/elasticsearch/core/internal/net/NetUtils.java

```diff
@@ -59,4 +64,45 @@
             return null;
         }
     }
+
+    /**
+     * If SO_KEEPALIVE is enabled (default), this method ensures sane default values for the extended socket options
+     * TCP_KEEPIDLE and TCP_KEEPINTERVAL. The default value for TCP_KEEPIDLE is system dependent, but is typically 2 hours.
+     * Such a high value can result in firewalls eagerly closing these connections. To tell any intermediate devices that
+     * the connection remains alive, we explicitly set these options to 5 minutes if the defaults are higher than that.
+     */
+    public static void setSaneDefaultKeepAliveOptions(NetworkChannel socketChannel) {
+        assert socketChannel != null;
+        try {
+            if (socketChannel.supportedOptions().contains(StandardSocketOptions.SO_KEEPALIVE)) {
+                final Boolean keepalive = socketChannel.getOption(StandardSocketOptions.SO_KEEPALIVE);
+                if (keepalive != null && keepalive.booleanValue()) {
+                    for (SocketOption<Integer> option : Arrays.asList(
+                        NetUtils.getTcpKeepIdleSocketOptionOrNull(),
+                        NetUtils.getTcpKeepIntervalSocketOptionOrNull())) {
+                        setMinValueForSocketOption(socketChannel, option, 300);
+                    }
+                }
+            }
+        } catch (Exception e) {
+            assert e instanceof IOException || e instanceof ClosedChannelException :
+                "unexpected exception when setting channel option: " + e.getClass() + ": " + e.getMessage();
+        }
+    }
+
+    private static void setMinValueForSocketOption(NetworkChannel socketChannel, SocketOption<Integer> option, int minValue) {
+        if (option != null && socketChannel.supportedOptions().contains(option)) {
+            try {
+                final Integer currentIdleVal = socketChannel.getOption(option);
+                if (currentIdleVal != null && currentIdleVal.intValue() > minValue) {
```

Comment:
I understand that you are being safe, but is there a known scenario where the socket option is supported, but the current value is null?

Reply:
I was concerned by the javadocs on the `getOption` method which says:

```
@return  The value of the socket option. A value of {@code null} may be a valid value for some socket options.
```

Rereading the Javadocs for the various socket options, it should be clear from their javadocs that the ones used are not returning null. I've amended the logic accordingly (but added an assertion instead)